File: ds_7_array_inplace/L565-m.cpp

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int arrayNesting(vector<int>& nums) {
        int res = 0;
        for(int i=0; i< nums.size(); ++i){
            int tmp_len = 0,index = i;
            //当前等于-1说明之前的元素已经访问过当前位置了
            while(nums[index] != -1){
                ++tmp_len;
                int tmp = nums[index];
                cout << nums[index]<<endl;
                nums[index] = -1; // 标记该位置已经被访问
                index = tmp;
            }
            cout << "**************" << endl;
            res = max(res, tmp_len);
        }
        return res;
    }

};

//超时，上面的做法则是做了剪枝
class Solution1 {
public:
    int arrayNesting(vector<int>& nums) {
        int res = 0;
        for(int i=0; i< nums.size(); ++i){
            int tmp_len = 0,index = i;
            do{
                ++tmp_len;
                cout << nums[index]<<endl;
                index = nums[index];
            }while(index != i);
            cout << "**************" << endl;
            res = max(res, tmp_len);
        }
        return res;
    }

};

int main(){
    vector<int> v1 = {5,4,0,3,1,6,2};
    int res = Solution().arrayNesting(v1);
    //cout << res << endl;
}

File: python-algorithm/ds_7_array_inplace/L565-m.py
# -*- coding: utf-8 -*-


# 提示：
# N是[1, 20,000]之间的整数。
# A中不含有重复的元素。
# A中的元素大小在[0, N-1]之间。

# 剪枝很重要，否则超时
class Solution:
    def arrayNesting(self, nums: List[int]) -> int:
        res = 1
        visited = [0 for i in range(len(nums))]
        for i in range(len(nums)):
            if not visited[i]:
                visited[i] = 1
                cnt = 1
                nxt = nums[nums[i]]
                while nxt != nums[i]:
                    visited[nxt] = 1
                    cnt += 1
                    nxt = nums[nxt]
                res = max(res, cnt)
        return res

#    更省空间的写法
class Solution:
    def arrayNesting(self, nums: List[int]) -> int:
        res = 1
        for i in range(len(nums)):
            tmp_len = 0
            idx = i
            while nums[idx] != -1:
                tmp_len += 1
                nxt_idx = nums[idx]
                # 访问过的被标记成-1
                nums[idx] = -1
                idx = nxt_idx
            res = max(res, tmp_len)
        return res